barretenberg/cpp/src/barretenberg/vm/avm_trace/AvmMini_execution.cpp

```diff
@@ -0,0 +1,227 @@
+                throw std::runtime_error("Instruction tag for SET opcode is invalid at position " +
+                                         std::to_string(pos) + " value: " + std::to_string(in_tag_u8));
+                break;
+            }
+        } else {
+            num_of_operands = Bytecode::OPERANDS_NUM.at(opcode);
+            operands_size = AVM_OPERAND_BYTE_LENGTH * num_of_operands;
+        }
+
+        if (pos + operands_size > length) {
+            throw std::runtime_error("Operand is missing at position " + std::to_string(pos));
+        }
+
+        if (opcode == OpCode::SET && in_tag == AvmMemoryTag::U8) {
+            operands.push_back(static_cast<uint32_t>(bytecode.at(pos)));
+            uint8_t const* ptr = &bytecode.at(pos + 1);
+            uint32_t operand{};
+            serialize::read(ptr, operand);
+            operands.push_back(operand);
+            pos += operands_size;
+        } else if (opcode == OpCode::SET && in_tag == AvmMemoryTag::U16) {
+            uint8_t const* ptr = &bytecode.at(pos);
+            uint16_t operand{};
+            serialize::read(ptr, operand);
+            operands.push_back(static_cast<uint32_t>(operand));
+            ptr = &bytecode.at(pos + 2);
+            uint32_t operand2{};
+            serialize::read(ptr, operand2);
+            operands.push_back(operand2);
+            pos += operands_size;
+        } else {
+            for (size_t i = 0; i < num_of_operands; i++) {
+                uint8_t const* ptr = &bytecode.at(pos);
+                uint32_t operand{};
+                serialize::read(ptr, operand);
+                operands.push_back(operand);
+                pos += AVM_OPERAND_BYTE_LENGTH;
+            }
+        }
+
+        instructions.emplace_back(opcode, operands, static_cast<AvmMemoryTag>(in_tag));
+    }
+
+    return instructions;
+}
+
+/**
+ * @brief Generate the execution trace pertaining to the supplied instructions.
+ *
+ * @param instructions A vector of the instructions to be executed.
+ * @param calldata expressed as a vector of finite field elements.
+ * @return The trace as a vector of Row.
+ */
+std::vector<Row> Execution::gen_trace(std::vector<Instruction> const& instructions, std::vector<FF> const& calldata)
+{
+    AvmMiniTraceBuilder trace_builder{};
+
+    for (auto const& inst : instructions) {
+        switch (inst.op_code) {
+        case OpCode::ADD:
+            trace_builder.add(inst.operands.at(0), inst.operands.at(1), inst.operands.at(2), inst.in_tag);
+            break;
+        case OpCode::SUB:
+            trace_builder.sub(inst.operands.at(0), inst.operands.at(1), inst.operands.at(2), inst.in_tag);
+            break;
+        case OpCode::MUL:
+            trace_builder.mul(inst.operands.at(0), inst.operands.at(1), inst.operands.at(2), inst.in_tag);
+            break;
+        case OpCode::DIV:
+            trace_builder.div(inst.operands.at(0), inst.operands.at(1), inst.operands.at(2), inst.in_tag);
+            break;
```

Comment:
Probably best to think about how we've implemented the TS instruction classes and the TS "interpreter" and keep track of our decisions to do things the same/differently here.

I'm perfectly fine with the answer "let's just make it a switch-case to keep it simple for now", or "this way of representing instructions and this switch-case may be more efficient than the more object-oriented/functional way it's handled in TS." But it'd be good to get it written in comments or elsewhere just so in the future it's easier to decide to change it not.